File: services/csfloat.go
package services

import (
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"github.com/cyberbebebe/cs2-profit-checker/types"
	"github.com/cyberbebebe/cs2-profit-checker/utils"
)

type CSFloatService struct {
	ApiKey string
	Client *http.Client
}

func NewCSFloatService(apiKey string) *CSFloatService{
	return &CSFloatService{
		ApiKey: apiKey,
		Client: &http.Client{Timeout: 10 * time.Second},
	}
}

func (s *CSFloatService) Name() string {
	return "CSFloat"
}

// SALE HISTORY
func (s *CSFloatService) GetSales(startTime, endTime time.Time)([]types.Transaction, error){
	return s.fetchTrades("seller", startTime, endTime)
}

// BUY HISTORY
func (s *CSFloatService) GetBuys(startTime, endTime time.Time)([]types.Transaction, error){
	return s.fetchTrades("buyer", startTime, endTime)
}

// Shared fetcher
func (s *CSFloatService) fetchTrades(role string, startTime, endTime time.Time) ([]types.Transaction, error){
	var allTxs []types.Transaction
	page := 0

	for{
		url := fmt.Sprintf("https://csfloat.com/api/v1/me/trades?role=%s&state=verified&limit=1000&page=%d", role, page)

		req, _ := http.NewRequest("GET", url, nil)

		req.Header.Set("Authorization", s.ApiKey)

		resp, err := s.Client.Do(req)
		if err != nil{
			return nil, fmt.Errorf("request failed: %v", err)
		}
		defer resp.Body.Close()

		if resp.StatusCode != 200{
			return nil, fmt.Errorf("CSFloat API Error: %d", resp.StatusCode)
		}

		var response csfloatResponse
		if err := json.NewDecoder(resp.Body).Decode(&response); err != nil{
			return nil, fmt.Errorf("decode error: %v", err)
		}

		if len(response.Trades) == 0 {
			break
		}

		for _, raw := range response.Trades{
			cleanTx := convertCsfloatTx(raw, role)

			if cleanTx.Date.After(endTime) {
				continue
			}
			if cleanTx.Date.Before(startTime) {
				return allTxs, nil
			}

			allTxs = append(allTxs, cleanTx)
		} 

		page++
		time.Sleep(1500 * time.Millisecond)
	}
	return allTxs, nil
}

func convertCsfloatTx(raw trade, role string) types.Transaction{
	t, err := time.Parse(time.RFC3339, raw.VerifiedAt)
	if err != nil {
		t = time.Now() // Fallback
	}

	txType := types.TxSell
	if role == "buyer" {
		txType = types.TxBuy
	}

	var finalPrice float64

	// Calculate 98% of the value if our sale
	if role == "seller" {
        afterFeeCents := float64(raw.Contract.Price) * 0.98
        finalPrice = utils.FloatRound((afterFeeCents) / 100.0)

    } else {
        finalPrice = float64(raw.Contract.Price) / 100.0
	}

	pattern := raw.Contract.Item.Pattern

	if raw.Contract.Item.CharmPattern != 0{
		pattern = raw.Contract.Item.CharmPattern
	}

	tx := types.Transaction{
		Source:    "CSFloat",
		Type:      txType,
		TxID:      raw.ID,
		AssetID:   raw.Contract.Item.AssetID,
		ItemName:  raw.Contract.Item.ItemName,
		Price:     finalPrice,
		Currency:  "USD",
		Date:      t,
		FloatVal:  raw.Contract.Item.FloatValue,
		Pattern:   pattern,
		Phase:     raw.Contract.Item.Phase,
	}

	tx.Signature = utils.GenerateSignature(tx.ItemName, tx.FloatVal, tx.Pattern)
	
	return tx

}

File: js/fetchers/csfloat.js
import { BaseFetcher } from "./base.js";
import { Transaction } from "./base.js";

export class CSFloatFetcher extends BaseFetcher {
  constructor() {
    super("CSFloat");
  }

  async checkSession() {
    try {
      await this.fetchWithAuth("https://csfloat.com/api/v1/me");
      return true;
    } catch (e) {
      return false;
    }
  }
  async getBalance() {
    try {
      const url = "https://csfloat.com/api/v1/me";
      const data = await this.fetchWithAuth(url);

      if (!data) return { amount: 0, currency: "USD" };

      const usd = parseFloat(data.user.balance || 0);
      const locked = parseFloat(data.user.pending_balance || 0);

      const total = (usd + locked) / 100.0;
      return { amount: total, currency: "USD" };
    } catch (e) {
      console.error("[CSFloat] Balance error:", e);
      return { amount: 0, currency: "USD" };
    }
  }

  async getHistory(role) {
    let items = [];
    let page = 0;
    const limit = 1000;

    while (true) {
      const url = `https://csfloat.com/api/v1/me/trades?role=${role}&state=pending,verified&limit=${limit}&page=${page}`;

      let data;
      try {
        data = await this.fetchWithAuth(url);
      } catch (e) {
        break;
      }

      if (!data || !data.trades || data.trades.length === 0) break;

      for (const trade of data.trades) {
        const item = trade.contract.item;
        const rawPrice = trade.contract.price / 100.0;
        let finalPrice = rawPrice;
        const verifiedDate = trade.verified_at
          ? new Date(trade.verified_at)
          : null;
        const createdDate = trade.created_at
          ? new Date(trade.created_at)
          : verifiedDate || new Date();
        if (role === "seller") {
          finalPrice = rawPrice * 0.98;
        }

        items.push(
          new Transaction({
            source: "CSFloat",
            type: role === "seller" ? "SELL" : "BUY",
            tx_id: trade.id,
            asset_id: item.asset_id,
            item_name: item.market_hash_name,
            price: parseFloat(finalPrice.toFixed(2)),
            currency: "USD",
            created_at: createdDate,
            verified_at: verifiedDate,
            float_val: item.float_value,
            pattern: item.paint_seed,
            phase: item.phase,
          }),
        );
      }

      if (data.trades.length < limit) break;
      page++;
      await this.sleep(1000);
    }
    return items;
  }

  async getSales() {
    return this.getHistory("seller");
  }
  async getBuys() {
    return this.getHistory("buyer");
  }
}
